src/core/queriesObserver.ts

```diff
@@ -64,57 +64,94 @@ export class QueriesObserver extends Subscribable<QueriesObserverListener> {
   }
 
   getOptimisticResult(queries: QueryObserverOptions[]): QueryObserverResult[] {
-    return queries.map((options, index) => {
-      const defaultedOptions = this.client.defaultQueryObserverOptions(options)
-      return this.getObserver(defaultedOptions, index).getOptimisticResult(
-        defaultedOptions
-      )
-    })
+    return this.findMatchingObservers(queries).map(match =>
+      match.observer.getCurrentResult()
+    )
   }
 
-  private getObserver(
-    options: QueryObserverOptions,
-    index: number
-  ): QueryObserver {
+  private findMatchingObservers(
+    queries: QueryObserverOptions[]
+  ): QueryObserverMatch[] {
+    const prevObservers = this.observers
+    const defaultedQueryOptions = queries.map(options =>
+      this.client.defaultQueryObserverOptions(options)
+    )
+
+    const matchingObservers: QueryObserverMatch[] = defaultedQueryOptions
+      .map(defaultedOptions => {
+        const match = prevObservers.find(
+          observer => observer.options.queryHash === defaultedOptions.queryHash!
+        )
+        if (match != null) {
+          return { defaultedQueryOptions: defaultedOptions, observer: match }
+        }
+        return null
+      })
+      .filter(notNullOrUndefined)
```

Comment:
I think we wouldn't need the `notNullOrUndefined` util if we'd just `flatMap` over the array and return `[]` for cases that we want to exclude and an array with a single entry for those that match. What do you think?

Reply:
Good point, `flatMap` would work here as well, and then we would not need the new util for the single use.

Replaced now with `flatMap` 👍